### PR TITLE
Differentiate between datetimeformat and dateformat

### PR DESCRIPTION
**Pull request by knagra**
Marshmallow currently uses `dateformat` in the `DateTime` field and does not support a custom format for `Date`. This PR changes `DateTime.dateformat` to `DateTime.DateTimeFormat` and adds `Date.dateformat`.

This is my first PR for this repo. Please let me know if I'm breaking any protocols or if other changes are necessary for this to be merged.

### CODE REVIEW

File: marshmallow/fields.py [hunk 1]
```diff
@@ -1016,32 +1027,86 @@ def _deserialize(self, value, attr, data):
 class Date(Field):
     """ISO8601-formatted date string.
 
+    :param strict: If True, only accept Date objets. If False, will accept any
```
Comment (sloria) [inline marshmallow/fields.py]: Drive-by: There is some discussion about removing `strict` flags on fields here: https://github.com/marshmallow-code/marshmallow/pull/755#issuecomment-377662237 , so I have reservations about adding it to `Date`.

**Comment by sloria** [comment]
Thanks @knagra for the PR. I think you're onto something here. I don't have time to review this deeply right now, but that's not to say it won't get reviewed/merged. I'm currently focused on moving marshmallow 3 forward, which means reviewing backwards-incompatible API changes.

If anyone else would like to review this, all feedback is welcome.

**Comment by knagra** [comment]
I can remove the `strict` flag. Also, this is a backwards-incompatible change. Currently, `DateTimeField` uses `dateformat` in `meta`. Under these changes, `DateTimeField` will use `datetimeformat` in `meta` and `DateField` will use `dateformat`. Thanks for the quick initial look, @sloria.

**Comment by sloria** [comment]
Yeah, go ahead and remove the `strict` option for now. That can be discussed separate from this PR. I'm good with the overall approach here. Since this is a breaking change, I'd like at least one more review of this.  cc @deckar01 @lafrech

File: marshmallow/fields.py [hunk 2]
```diff
@@ -1016,32 +1027,86 @@ def _deserialize(self, value, attr, data):
...
+            self.fail('format', input=value)
         try:
-            return value.isoformat()
-        except AttributeError:
+            return dt.date(value.year, value.month, value.day)
+        except (TypeError, AttributeError, ValueError):
             self.fail('format', input=value)
-        return value
+
+    def _serialize(self, value, attr, obj):
```
Comment (deckar01) [inline marshmallow/fields.py]: A lot of this new code seems to be duplicated from `DateTime`. Can we use inheritance to avoid duplicating this code?

File: marshmallow/fields.py [hunk 3]
```diff
@@ -1016,32 +1027,86 @@ def _deserialize(self, value, attr, data):
...
+        'iso': utils.from_iso_date,
+        'iso8601': utils.from_iso_date,
+    }
+
+    DEFAULT_FORMAT = 'iso'
+
+    def __init__(self, strict=False, format=None, **kwargs):
+        super(Date, self).__init__(**kwargs)
+        self.strict = strict
+        self.dateformat = format
```
Comment (deckar01) [inline marshmallow/fields.py]: It might be better to avoid field name specific prefixes to make these properties nicer to override with inheritance. They are already scoped under a class, so there is no benefit to further namespacing them.

- `self.dateformat` -> `self.format`
- `DATEFORMAT_SERIALIZATION_FUNCS` -> `SERIALIZATION_FUNCS`
- `DATEFORMAT_DESERIALIZATION_FUNCS ` -> `DESERIALIZATION_FUNCS `

File: marshmallow/utils.py [hunk 4]
```diff
@@ -309,6 +309,11 @@ def from_iso_date(datestring, use_dateutil=True):
     else:
         return datetime.datetime.strptime(datestring[:10], '%Y-%m-%d').date()
 
+
+def to_iso_date(date):
+    return date.strftime('%Y-%m-%d')
```
Comment (fuhrysteve) [inline marshmallow/utils.py]: This functionality already exists in the stdlib as `datetime.date.isoformat` - so we should be able to use that instead of adding another utils function.

```python
In [1]: import datetime

In [2]: datetime.date.today().isoformat()
Out[2]: '2018-07-12'

In [3]: datetime.date.isoformat(datetime.date.today())
Out[3]: '2018-07-12'
```

Comment (knagra) [reply marshmallow/utils.py]: I'm fine with that. I'll make the change.

File: marshmallow/fields.py (hunk 3 above)
Comment (knagra) [reply marshmallow/fields.py]: That makes sense to me. I'll make the changes.

File: marshmallow/fields.py (hunk 2 above)
Comment (knagra) [reply marshmallow/fields.py]: The only major impediment is the distinction between `dateformat` and `datetimeformat` in Meta, but I think I can write a function to change that between the two classes.

How are you envisioning the class inheritance graph for this? `class Date(DateTime)` or `class Date(FormattedField)` and `class DateTime(FormattedField)` or something similar?

Comment (deckar01) [reply marshmallow/fields.py]: I don't think python has a defined hierarchy between the two modules. I think it would be valid to treat `Date` as a specialized form of `DateTime`. `class DateTime(Field)` and `class Date(DateTime)` would probably generate the smallest diff also.

> I can write a function to change [the meta format] between the two classes

:+1: Good catch.

**Comment by lafrech** [comment]
@knagra thanks for this work, already. Are you still working on this?

**Comment by knagra** [comment]
Yeah, I'm about to work on this today and tomorrow. Just been busy with work.

**Comment by lafrech** [comment]
No pb, no pressure. Thanks.

**Comment by knagra** [comment]
@deckar01 @fuhrysteve I've incorporated your suggested changes. Implementing `Date(DateTime)` inheritance required adding a couple of extra class variables to those fields to maintain cleanliness of error messages and to allow schemas to have both `dateformat` and `datetimeformat` in the `Meta` class.

I converted the `to_iso_format` function to use `datetime.date.isoformat` as suggested but decided to leave to function to absorb the extra `self.localtime` keyword argument passed to it by `DateTime`.

Let me know if you have other suggestions. I should be able to address them more quickly now.

File: marshmallow/fields.py [hunk 5]
```diff
@@ -931,62 +931,74 @@ class DateTime(Field):
...
 
     def _add_to_schema(self, field_name, schema):
         super(DateTime, self)._add_to_schema(field_name, schema)
-        self.dateformat = self.dateformat or schema.opts.dateformat
+        self.format = self.format or \
+            getattr(schema.opts, self.SCHEMA_OPTS_VAR_NAME) or \
+            self.DEFAULT_FORMAT
 
     def _serialize(self, value, attr, obj):
+        self.format = self.format or self.DEFAULT_FORMAT
```
Comment (lafrech) [inline marshmallow/fields.py]: This will conflict with https://github.com/marshmallow-code/marshmallow/pull/934.

Please don't mutate `self.format` and use a local `format` variable.

Comment (knagra) [reply marshmallow/fields.py]: Okay. It seems that change was merged just a few days ago. I'll go ahead and change `self.format` to just `format` in this case. Do you want me to change the name of the `self.format` variable back to `self.dateformat` as well? I changed it since `dateformat` in the `DateTime` field doesn't make sense.

File: tests/test_schema.py [hunk 6]
```diff
@@ -1468,19 +1468,32 @@ class BadExclude(Schema):
...
             fields = ('created', 'updated')
-            dateformat = fmt
-    serialized = DateFormatSchema().dump(user)
+            datetimeformat = fmt
+    serialized = DateTimeFormatSchema().dump(user)
     assert serialized['created'] == user.created.strftime(fmt)
     assert serialized['updated'] == user.updated.strftime('%m-%d')
 
+def test_dateformat_option(user):
+    fmt = '%Y-%m'
```
Comment (lafrech) [inline tests/test_schema.py]: Very minor: it is strange to define this format here and no the other one.

File: marshmallow/fields.py (hunk 5 above)
Comment (lafrech) [reply marshmallow/fields.py]: > It seems that change was merged just a few days ago. I'll go ahead and change self.format to just format in this case.

Yes, exactly. I though I'd mention that to help you with the rebase.

> Do you want me to change the name of the self.format variable back to self.dateformat as well?

No, not at all. `format` is better now.

File: tests/test_schema.py (hunk 6 above)
Comment (knagra) [reply tests/test_schema.py]: It was the pattern saw I above at line 1471. I'll go ahead name them in both tests.

File: marshmallow/fields.py [hunk 7]
```diff
@@ -931,62 +931,74 @@ class DateTime(Field):
...
             except (TypeError, AttributeError, ValueError):
-                raise self.fail('invalid')
+                raise self.fail('invalid', obj_type=self.OBJ_TYPE)
         elif utils.dateutil_available:
             try:
-                return utils.from_datestring(value)
-            except TypeError:
-                raise self.fail('invalid')
+                parsed = utils.from_datestring(value)
+                return dt.datetime(
```
Comment (lafrech) [inline marshmallow/fields.py]: I don't understand this change. I don't think this would return a `dt.date` when called from `Date`.

Isn't this code unreachable anyway ? (https://github.com/marshmallow-code/marshmallow/issues/758)

(I don't mean to throw #758 in your way, this is another story. If the code is indeed unreachable, which I think it is, don't worry too much about it. I just don't understand the change.)

File: tests/test_schema.py (hunk 6 above)
Comment (lafrech) [reply tests/test_schema.py]: Oh, right, of course. Thanks. No pb otherwise.

File: marshmallow/fields.py (hunk 5 above)
Comment (lafrech) [reply marshmallow/fields.py]: Sorry, I just realized this PR is rebased already. Then please change `self.format` to `format` as in #934.

File: marshmallow/fields.py (hunk 7 above)
Comment (knagra) [reply marshmallow/fields.py]: The object creation is an oversight. I'll fix that. I think #758 is beyond the scope of this PR. I can look at it after this change, though, if you need someone to work on it.

**Comment by knagra** [comment]
@lafrech Thanks for the review. I'll implement changes tonight and have it ready for another take.

**Comment by knagra** [comment]
@lafrech I implemented your suggestions. Yes, the code block you mentioned is inaccessible. That's why we weren't getting any errors from the tests. I don't think there's a clear quick fix for #758. There are some behavior questions that need to be answered.

**Comment by lafrech** [comment]
Thanks @knagra for your work on this.

I think this is ready to merge.

**Comment by lafrech** [comment]
@knagra, if you have time, would you mind adding a paragraph to docs/upgrading.rst?

**Comment by lafrech** [comment]
I merged and updated the docs.

Thanks again @knagra. Please double-check the doc update if you get the time.
